### PR TITLE
Add examples to transform docs

### LINKED ISSUE #1409: Add examples of transformation to docs
# Problem
`torchvision` has quite a few transforms, but I often find myself using the wrong one because the I misread/just skimmed the doc strings.

# Solution
I'm suggesting to add examples / visualizations for each transform. This way it's easier to know what a given transform does.

After this improvement there should be an example (maybe a few where it makes sense) for each transform in the docs:
https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision-transforms

Also, a picture is worth a thousand words ;)

What do you think?

### PR DESCRIPTION
**Pull request by sotte**
This PR is the result of https://github.com/pytorch/vision/issues/1409. @fmassa thanks for the input!

The PR adds examples of each transform class to the sphinx docs. The result looks something like this:
image

## Make `_sample_image()` public
@fmassa said:
> Also, the Grace Hopper image is available under test/assets I believe, which is not installed with torchvision, so this would mean that we would need to move the image somewhere else. My thinking is that those two functions are very specific to the examples, and thus having them live in torchvision.utils as being public functions might not be best.

I would argue that it's quite convenient to be able to get sample image and therefore `sample_image()` should be public. [skimage](https://github.com/scikit-image/scikit-image/blob/master/skimage/data/__init__.py) for example gives you quite a few sample images which I actually tend to use when using pytorch.
(That being said I'm also fine with a private `_sample_image()`.)

## Implementation of `_sample_image()` public
The current implementation is just a dummy implementation that load the image from /tmp :)
The Hopper image is part of the `test/` folder and is not part of the resulting egg/wheel/whatever.
Normally I would just use https://docs.python.org/3.7/library/importlib.html#module-importlib.resources to load the file, but pytorch is not 3.7 only :) So we have to decide on where to put the file and how to load it. I'm no expert when it comes to packaging and I'm more than open for suggestion.

## TODOs
- [x] Add examples to transform classes (as suggested)
- [x] Add helpers to load and display images
- [x] Split transform.rst into transform.rst and transform_functional.rst (as suggested)
- [x] Decide if `_sample_image()` really should de private
- [x] Add proper implementation for `_sample_image()` depending on how the actual sample image is going to be shipped.
- [x] Fix inconsistencies of of documented transforms.
- [x] Document all `__call__` of TensorTransforms.
- [ ] Clean git history.

Feedback welcome!

Closes #1409

### CODE REVIEW

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -165,13 +165,7 @@ def __init__(self, mean, std, inplace=False):
         self.inplace = inplace
 
     def __call__(self, tensor):
-        """
-        Args:
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Any particular reason why you removed this part of the docstring?

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -819,13 +813,7 @@ def __init__(self, transformation_matrix, mean_vector):
         self.mean_vector = mean_vector
 
     def __call__(self, tensor):
-        """
-        Args:
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: So you think it's better to remove those parts of the docstrings? It still carry some useful information, like the expected shapes of the tensors.

File: torchvision/transforms/transforms.py [hunk 3]
```diff
@@ -1225,15 +1215,13 @@ class RandomErasing(object):
...
-        Erased Image.
-    # Examples:
+    Example:
         >>> transform = transforms.Compose([
-        >>> transforms.RandomHorizontalFlip(),
-        >>> transforms.ToTensor(),
-        >>> transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
-        >>> transforms.RandomErasing(),
-        >>> ])
+        ...     transforms.RandomHorizontalFlip(),
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Does this work if we mix `>>>` and `...` ?

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -1288,13 +1276,7 @@ def get_params(img, scale, ratio, value=0):
         return 0, 0, img_h, img_w, img
 
     def __call__(self, img):
-        """
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: Same comment about the docstring

File: docs/source/transforms.rst [hunk 5]
```diff
@@ -9,124 +9,304 @@ Functional transforms give fine-grained control over the transformations.
...
+
 .. autoclass:: ColorJitter
 
+**Example**
+
+.. plot::
+   :include-source:
+
+    from torchvision.transforms import ColorJitter
+    from torchvision.utils import _plot_images, _sample_image
```
Comment (fmassa) [inline docs/source/transforms.rst]: I was thinking if it wouldn't be maybe better to have those helper functions be in the documentation folder, instead of torchvision?

I don't think we want the docs to show private functions, as they would in fact encourage people to use those private functions

File: torchvision/transforms/transforms.py (hunk 2 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: You're right, the shape might be of interest.

File: torchvision/transforms/transforms.py (hunk 3 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Yes, this is actually how doctests should be written. This also mimics the python REPL.

(Note that the indentation was of in the earlier version.)

Just to verify that it's valid here is a minimal example:
```python
class Foo:
    """Foooo

    Examples:
        >>> foo = Foo(
        ...     name="Alice"
        ... )
        >>> foo.foo()
        Alice
    """
    def __init__(self, name):
        self.name = name

    def foo(self):
        print(self.name)
```
Then
```bash
pytest --doctest-modules dummy.py
```

File: docs/source/transforms.rst (hunk 5 above)
Comment (sotte) [reply docs/source/transforms.rst]: My thought is that it would be nice to be able to copy and paste the examples and go from there. If the two functions are only part of the docs folder then it's not possible to reproduce the example.

I'm not sure how to best add these function only to the sphinx build.

File: torchvision/transforms/transforms.py (hunk 1 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Just to underline my point here are the three different ways `__call__` is currently documented (or not):

image
image

Now it looks like this:
image

Comment (fmassa) [reply torchvision/transforms/transforms.py]: Ok with me, but I'd like to keep the input / output shape information if possible

File: docs/source/transforms.rst (hunk 5 above)
Comment (fmassa) [reply docs/source/transforms.rst]: Hum, I'm not sure what's the best way to go here.

Asking for feedback from @brianjo @soumith if we should have those helper functions in torchvision or not, given that they will be mostly used for illustration purposes.

**Comment by fmassa** [comment]
Another thing, we should not forget to fix the random seeds (for both `random` and `torch`) to avoid the images being regenerated at every doc rebuild.

File: torchvision/transforms/transforms.py (hunk 1 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Should be addressed in b9a76785c382fac762827ba5a945d03cd56c3711

File: torchvision/transforms/transforms.py (hunk 2 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Should be addressed in b9a76785c382fac762827ba5a945d03cd56c3711

File: torchvision/transforms/transforms.py (hunk 4 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Should be addressed in b9a76785c382fac762827ba5a945d03cd56c3711

**Comment by sotte** [comment]
> Another thing, we should not forget to fix the random seeds (for both `random` and `torch`) to avoid the images being regenerated at every doc rebuild.

The examples are only re-executed if the example code changes, i.e. calling `make html` twice in a row does not rebuild the images. Therefore I'm not not sure if he even have to fix the seeds. @fmassa if only recreating the images on every `make html` is the issue, then we don't need to fix the seeds. What do you think.

File: docs/source/transforms.rst (hunk 5 above)
Comment (sotte) [reply docs/source/transforms.rst]: Ok, waiting for input here.

**Comment by sotte** [comment]
We need to decide on the actual implementation of `_example_image()`, where the image is stored, and how the image should be packed into the wheel.

Just as reference: this is how skimage does it:
- Packing: https://github.com/scikit-image/scikit-image/blob/master/setup.py#L160
- Loading the images: https://github.com/scikit-image/scikit-image/blob/master/skimage/data/__init__.py#L78

File: docs/source/transforms.rst (hunk 5 above)
Comment (fmassa) [reply docs/source/transforms.rst]: pinging @jlin27 what do you think the above?

**Comment by fmassa** [comment]
> The examples are only re-executed if the example code changes, i.e. calling make html twice in a row does not rebuild the images

I might not understand how matplotlib caching works, but I would have expected that if I clean up the build folder and re-run again `make html` the images generated would be different? Which would mean the html would change everytime we update the documentation.

> We need to decide on the actual implementation of _example_image(), where the image is stored, and how the image should be packed into the wheel.

thanks for the pointers. I'm not yet clear if we want to have those functions in the torchvision distribution, but the pointers are anyway very helpful.

I'm waiting on feedback from @jlin27 on what would be better.

**Comment by jlin27** [comment]
Hi @fmassa @sotte 

My vote is to keep _sample_image() private, given as you've all discussed, that it's specific for this example use case. Helps signal to users that they probably shouldn't be using it outside of this situation.

As for where to store the images, I'd suggest a static folder with a descriptive name. For example, in pytorch/tutorials, images are stored  /static/_img (https://github.com/pytorch/tutorials/tree/master/_static/images), and just make sure to double-check that they are pulled in properly during build.

**Comment by sotte** [comment]
**Seeds**
@fmassa I think we were talking about different things. When you remove the build/ folder of the docs, of course the docs will be rebuild from scratch and the output might be different depending on the RNG. If you don't remove the build/ folder then sphinx is not going to re-execute the examples.
I set the random seed now and the examples should stay constant between complete rebuilds.

**Private functions**
I explicitly mention that the functions are not to be used.

**Loading and packaging the image / things that should be simple but are quite tricky**
@jlin27 thanks for the input! I think bundling the images is a bit more complex as in the pytorch tutorials case. We have to ship the assets.
In general I would add the image to a newly created `torchvision.assets` module so that it gets shipped with torchvision and so that I can access the file similar to how it's done with skimage. Something like this:
```python
def _sample_image():
    """Private helper function to load a sample PIL image.

    This function might change and/or break. Don't depend on it.
    """
    import os.path
    from PIL import Image

    data_dir = os.path.abspath(os.path.dirname(__file__))
    return Image.open(os.path.join(data_dir, "assets", "grace_hopper_517x606.jpg"))
```

Then of course we have to actually include the folder
```
# MANIFEST.in
include README.rst
include LICENSE
include torchvision/assets/*

recursive-exclude * __pycache__
recursive-exclude * *.py[co]
```
and change the
```python
# setup.py
setup(
    ...
    zip_safe=False,  # not zip safe because we load the image via a path
    include_package_data=True,
)
```
but after doing so I get the following error:
```
➤ python setup.py install 
Building wheel torchvision-0.5.0a0+5108c1e
running install
running bdist_egg
running egg_info
writing torchvision.egg-info/PKG-INFO
writing dependency_links to torchvision.egg-info/dependency_links.txt
writing requirements to torchvision.egg-info/requires.txt
writing top-level names to torchvision.egg-info/top_level.txt
reading manifest file 'torchvision.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '__pycache__' found under directory '*'
warning: no previously-included files matching '*.py[co]' found under directory '*'
writing manifest file 'torchvision.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
copying torchvision/version.py -> build/lib.linux-x86_64-3.7/torchvision
error: Error: setup script specifies an absolute path:

    /home/stefan/coding/torchvision/test/test_models.cpp

setup() arguments must *always* be /-separated paths relative to the
setup.py directory, *never* absolute paths.
```
I'm also not sure if we break things by making torchvision not `zip_safe`.

The packaging problem is still not solved and I'm more than open for input!


**Ref**
See the "note" box: https://python-packaging.readthedocs.io/en/latest/non-code-files.html

**Comment by fmassa** [comment]
Hi @sotte ,

Sorry for the delay in replying, I had to fix a few issues with torchvision lately (including making it `zip_safe=False` :-), see https://github.com/pytorch/vision/pull/1536 )

About the error you are facing, from a quick search it might indicate that you might need to do a clean build for torchvision?

Apart from that, I think that the approach you are following is the right one, with the `os.path.abspath(os.path.dirname(__file__))` and setting `include_package_data`.

I don't have much more experience with setuptools than that though

**Comment by sotte** [comment]
@fmassa no worries and thanks for the pointer. I'll give it a try and get back to you.

**Comment by sotte** [comment]
It's getting tricky :) Starting with a fresh build did not help.

Here is a log of what I did:
```
# Add the following to setup.py
#     zip_safe=False,
#     include_package_data=True,
python setup.py clean
python setup.py install
```
This yields the error:
```
...
warning: no previously-included files matching '__pycache__' found under directory '*'
warning: no previously-included files matching '*.py[co]' found under directory '*'
writing manifest file 'torchvision.egg-info/SOURCES.txt'
error: Error: setup script specifies an absolute path:

    /home/stefan/coding/torchvision/test/test_models.cpp

setup() arguments must *always* be /-separated paths relative to the
setup.py directory, *never* absolute paths.
```
Checking `torchvision.egg-info/SOURCES.txt` tells me that all *.cpp files are included with their absolute path:
```
$ cat torchvision.egg-info/SOURCES.txt | grep cpp                                                      
/home/stefan/coding/torchvision/test/test_models.cpp
/home/stefan/coding/torchvision/torchvision/csrc/vision.cpp
/home/stefan/coding/torchvision/torchvision/csrc/cpu/ROIAlign_cpu.cpp
/home/stefan/coding/torchvision/torchvision/csrc/cpu/ROIPool_cpu.cpp
/home/stefan/coding/torchvision/torchvision/csrc/cpu/nms_cpu.cpp
...
```

`get_extensions()` in `setup.py` seems to create absolute paths (line 80). I haven't had time to look into it more thoroughly.

**Comment by sotte** [comment]
Just to verify that I did not break anything in my branch: on master cd17484 when adding `include_package_data=True` I get the same error when calling `python setup.py build`.

**Comment by sotte** [comment]
@fmassa So I had to change the `setup.py` quite a bit (and I don't think this should be necessary) to be able to build torchvision when `include_package_data=True`. Before my commit  7d6d38c the `SOURCES.txt` contained absolute paths which should not be the case, but seemingly only produces errors when `include_package_data=True`. I'm quite confident that I broke the build with that commit though :)

Please, anybody who is more knowledgeable with python packaging take a look at this!

That being said, the PR should be complete now. Feel free to build the docs yourself and let me know if it's ok.

**Comment by sotte** [comment]
@fmassa Please verify that it works on you end. If so I think we can merge this one (after rebasing).

**Comment by fmassa** [comment]
Hi @sotte 

Very sorry for the delay in coming back to you.

I'll get this reviewed and merged tomorrow, thanks a lot!

**Comment by sotte** [comment]
Absolutely no problem. Just let me know if you need anything else.

**Comment by sotte** [comment]
@fmassa *friendly ping* :)

Let me know what I can do to get this into masteri

**Comment by fmassa** [comment]
Very sorry for the delay @sotte! Was very busy working on a paper submission.

I'm putting some time aside to review this tomorrow, thanks for the patience!

File: docs/source/transforms.rst [hunk 6]
```diff
@@ -9,124 +9,363 @@ Functional transforms give fine-grained control over the transformations.
 This is useful if you have to build a more complex transformation pipeline
 (e.g. in the case of segmentation tasks).
 
+.. Note::
+    Most transform classese have an equivalent in :mod:`torchvision.transforms.functional`.
```
Comment (vincentqb) [inline docs/source/transforms.rst:13]: nit: classes

File: torchvision/transforms/transforms.py [hunk 7]
```diff
@@ -156,12 +156,8 @@ def __init__(self, mean, std, inplace=False):
         self.inplace = inplace
 
     def __call__(self, tensor):
-        """
-        Args:
-            tensor (Tensor): Tensor image of size (C, H, W) to be normalized.
-
-        Returns:
-            Tensor: Normalized Tensor image.
+        """Apply the transform to the given tensor (C x H x W) and return the
```
Comment (vincentqb) [inline torchvision/transforms/transforms.py:159]: nit: Please make changes that are not related to the particular goal of the PR in a separate PR.

File: torchvision/transforms/transforms.py [hunk 8]
```diff
@@ -813,12 +810,8 @@ def __init__(self, transformation_matrix, mean_vector):
         self.mean_vector = mean_vector
 
     def __call__(self, tensor):
-        """
-        Args:
-            tensor (Tensor): Tensor image of size (C, H, W) to be whitened.
-
-        Returns:
-            Tensor: Transformed image.
+        """Apply the transform to the given tensor (C x H x W) and return the
```
Comment (vincentqb) [inline torchvision/transforms/transforms.py:813]: same.

File: torchvision/transforms/transforms.py [hunk 9]
```diff
@@ -1284,12 +1278,8 @@ def get_params(img, scale, ratio, value=0):
         return 0, 0, img_h, img_w, img
 
     def __call__(self, img):
-        """
-        Args:
-            img (Tensor): Tensor image of size (C, H, W) to be erased.
-
-        Returns:
-            img (Tensor): Erased Tensor image.
+        """Apply the transform to the given tensor (C x H x W) and return the
```
Comment (vincentqb) [inline torchvision/transforms/transforms.py:1281]: same.

File: torchvision/transforms/transforms.py (hunk 7 above)
Comment (sotte) [reply torchvision/transforms/transforms.py:159]: Thanks for getting back to me. I'll address these right away.

File: torchvision/transforms/transforms.py (hunk 9 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Fixed with 17db875.

File: torchvision/transforms/transforms.py (hunk 8 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Fixed with 17db875.

File: torchvision/transforms/transforms.py (hunk 7 above)
Comment (sotte) [reply torchvision/transforms/transforms.py]: Fixed with 17db875.

**Comment by sotte** [comment]
@fmassa  I was not able to build when using `abspath` and (I think) I was getting error messages similar to the current ones of circleCI: https://app.circleci.com/pipelines/github/pytorch/vision/2750/workflows/978a424b-8d15-4271-9ff5-df6cc9fa1360/jobs/151904/steps
```
setup() arguments must *always* be /-separated paths relative to the
setup.py directory, *never* absolute paths.
```

**Comment by NicolasHug** [comment]
Closing since https://github.com/pytorch/vision/pull/3652 has been merged.

Thanks a lot for the proposal and for initiating the work on this PR @sotte !

I opened https://github.com/pytorch/vision/issues/3688 as a follow-up
